working/4200-metaobjects/feature-specification.md

```diff
@@ -0,0 +1,519 @@
+declaration. For example, if a class `A` declares a static method `foo`
+then we can call it using `A.foo(...)`, but we can not call it using
+`X.foo(...)` where `X` is a type variable, even in the case where the value
+of `X` at run time is `A`.
+
+Here is an example:
+
+```dart
+// Define the interface that we wish to support.
+abstract class Fooable<X> {
+  const Fooable(); // A metaobject must be const-able.
+  String foo(X x); // This is a gateway to a static method.
+}
+
+// The `static implements` clause specifies that a
+// metaobject for `A` supports the given interface.
+class A static implements Fooable<A> {
+  final String name;
+  A(this.name);
+  static String foo(A a) => "${a.name} fooing!";
+}
+
+// Ditto.
+class B static implements Fooable<B> {
+  final int size;
+  B(this.size);
+  static String foo(B b) => "B of size ${b.size} fooing!";
+}
+
+// Does not depend on `A` or `B`, but is still type safe.
+String showFoo<X static extends Fooable<X>>(X x) {
+  return X.foo(x);
+}
+
+void main() {
+  print(showFoo(A("MyA"));
+  print(showFoo(B(42)));
+}
+```
+
+This works because the type variable `X` is evaluated to an object when it
+is used as the receiver in an expression like `X.foo(x)`. The
+_static bound_ on `X` which is specified as `static extends Fooable<X>`
+provides a compile-time guarantee that the actual argument that `X` is
+bound to in any given invocation of `showFoo` will be a type such that
+when it is evaluated as an expression, the resulting metaobject will have a
+type which is a subtype of `Fooable<X>`. In particular, it has a `foo`
+method that has a positional parameter of type `X`, so we can safely call
+it as `X.foo(x)`.
+
+In main, it is statically ensured that the actual type argument of the two
+invocations of `showFoo` satisfy this requirement: In the first invocation
+the inferred type argument is `A`, and this is OK because
+`A static implements Fooable<A>`. Similarly for the second invocation with
+the actual type argument `B`.
+
+The semantics that actually makes the invocation of `foo` call the static
+method of `A` respectively `B` is established by an implicitly induced
+class for each class that has a `static implements` clause, known as the
+_metaobject class_. The metaobject class implements all members of the
+specified interface by forwarding to a static member or a constructor of
+the underlying class, mixin, mixin class, or enum declaration.
+
+For example:
+
+```dart
+class MetaobjectForA implements Type, Fooable<A> {
+  const MetaobjectForA();
+  String foo(A a) => A.foo(a);
+  bool operator ==(Object other) => ...;
+  int get hashCode => ...;
+}
+
+class MetaobjectForB implements Type, Fooable<B> {
+  const MetaobjectForB();
+  String foo(B b) => B.foo(b);
+  bool operator ==(Object other) => ...;
+  int get hashCode => ...;
+}
+```
+
+These classes are implicitly generated by the compiler and analyzer, so we
+can't refer to them in user code (the real name is probably something like
```

Comment:
The "so" part doesn't follow from the former part.
Also mixing abstraction levels. The classes are *introduced* by the class declaration that has a `static implements` clause. They're not generated by a compiler, they exist *in the program as such* by the languge semantics.

What compilers do is an implementation detail that nobody other than compiler writers should need to know.